toolchain/check/context.cpp

```diff
@@ -681,39 +763,43 @@ static auto ConvertTupleToArray(Context& context, SemIR::Node tuple_type,
     value_id = context.ConvertToValueOrReferenceExpression(value_id);
   }
 
+  Context::PendingBlock target_block(context);
+
   // Arrays are always initialized in-place. Tentatively allocate a temporary
   // as the destination for the array initialization.
-  auto return_slot_id = context.AddNode(
+  auto return_slot_id = target_block.AddNode(
       SemIR::Node::TemporaryStorage::Make(value.parse_node(), array_type_id));
 
   // Initialize each element of the array from the corresponding element of the
   // tuple.
   llvm::SmallVector<SemIR::NodeId> inits;
   inits.reserve(array_bound + 1);
   for (auto [i, src_type_id] : llvm::enumerate(tuple_elem_types)) {
+    Context::PendingBlock::DiscardUnusedNodesScope scope(target_block);
     // TODO: Add a new node kind for indexing an array at a constant index
     // so that we don't need an integer literal node here.
-    auto index_id = context.AddNode(SemIR::Node::IntegerLiteral::Make(
+    auto index_id = target_block.AddNode(SemIR::Node::IntegerLiteral::Make(
         value.parse_node(),
         context.CanonicalizeType(SemIR::NodeId::BuiltinIntegerType),
         context.semantics_ir().AddIntegerLiteral(llvm::APInt(32, i))));
-    auto target_id = context.AddNode(SemIR::Node::ArrayIndex::Make(
+    auto target_id = target_block.AddNode(SemIR::Node::ArrayIndex::Make(
         value.parse_node(), element_type_id, return_slot_id, index_id));
-    auto src_id =
-        !literal_elems.empty()
-            ? literal_elems[i]
-            : context.AddNode(SemIR::Node::TupleIndex::Make(
-                  value.parse_node(), src_type_id, value_id, index_id));
-    auto init_id =
-        context.InitializeAndFinalize(value.parse_node(), target_id, src_id);
+    auto src_id = !literal_elems.empty()
+                      ? literal_elems[i]
+                      : context.AddNode(SemIR::Node::TupleAccess::Make(
```

Comment:
Should this be in `target_block` as well?

```suggestion
                      : target_block.AddNode(SemIR::Node::TupleAccess::Make(
```

And if so -- this seems awfully brittle. Is there anything to check / ensure we get this right?

Reply:
Added a comment to explain.